### DISCUSSION TITLE
Lowdefy v3.20 - Lots of things

### DISCUSSION THREAD

**Discussion by SamTolmay** [lowdefy/lowdefy#768] [Announcement]
We are happy to release Lowdefy v3.20. This release contains a number of features requested by the community or that we have needed when building apps. We hope this release makes the experience of working with Lowdefy more enjoyable and enables you to build better things 🚀

#### Actions

- A `debounce` option has been added to events.
- An `async` option has been added to actions so that they are not awaited in the action chain.
- A new `ResetValidation` action has been added.
- A new `Throw` action has been added.
- A `back` option has been added to the `Link` action.
- The Lowdefy action functions can now be used inside the `JsAction` action.
- The `Validate` action can now take a list of regular expressions to match blocks to validate.
- Only blocks that have been validated now show the validation result.

#### Blocks

- The `List` block now has `direction`, `wrap`, and `scroll` properties.

- HTML is now supported in block properties that used to only take strings. The following blocks now have support for HTML properties:

  - `Alert`
  - `AutoComplete`
  - `Button`
  - `ButtonSelector`
  - `Card`
  - `CheckboxSelector`
  - `CircleColorSelector`
  - `Collapse`
  - `ColorSelector`
  - `CompactColorSelector`
  - `ConfirmModal`
  - `Descriptions`
  - `Divider`
  - `GithubColorSelector`
  - `Label`
  - `Message`
  - `Modal`
  - `MultipleSelector`
  - `Notification`
  - `Paragraph`
  - `RadioSelector`
  - `Result`
  - `Selector`
  - `SliderColorSelector`
  - `Statistic`
  - `SwatchesColorSelector`
  - `Title`
  - `Tooltip`
  - `TwitterColorSelector`

- The `showTotal` property in the `Pagination` block can now be a string or function.
- If the `onClose` event fails in the `ConfirmModal`, `Drawer`, and `Modal` blocks, the block no longer closes.
- The `ParagraphInput` and `TitleInput` blocks now render non-truthy values.

#### CLI and build

- The `_ref` operator can now specify a resolver function that overrides the default reading of configuration files from the file system.
- A default `_ref` resolver function can be specified for an app.
- The CLI can now be configured from the `lowdefy.yaml` file.
- The `dev` command now has `watch` and `watchIgnore` options to control which files are watched for rebuilds.
- The blocks server URL is now configurable.
- The `dev` server no longer exits if the initial build fails.

#### Connections

- The following Elasticsearch requests have been added:
  - `ElasticsearchDelete`
  - `ElasticsearchDeleteByQuery`
  - `ElasticsearchIndex`
  - `ElasticsearchUpdate`
  - `ElasticsearchUpdateByQuery`
- Read and write checking for Elasticsearch has been added.
- The `mongodb` driver has been updated to 4.1.0, and now supports different options.
- Connection options are now passed to the `MongoDBCollection` connection.

#### Docs

- The `AxiosHttp` examples have been fixed.
- The Netlify deployment steps have been updated.

#### Operators

- The `_location` operator now returns the `basePath`, `homePageId`, and `pageId`.
- The `_number` operator has been added.

#### Servers

- The server `basePath` is now configurable.
- A Node.js production server has been added.

<hr /><em>This discussion was created from the release <a href='https://github.com/lowdefy/lowdefy/releases/tag/v3.20.0'>v3.20.0</a>.</em>

**Comment by Radiergummi** [comment]
Awesome, super excited for this release! :heart:

**Reply by SamTolmay** [reply]
Thanks. And thank you for the work on Elasticsearch 😄
